### PR TITLE
[Gf] implement BlockGf variants of DLR functions

### PR DESCRIPTION
**Pull request by the-hampel**

### CODE REVIEW

File: test/c++/gfs/gf_dlr.cpp [hunk 1]
```diff
@@ -352,6 +352,15 @@ TEST(Gf, DLR_imtime_fit) {
     auto gtau3       = make_gf_imtime(gcoef_noise, n_tau);
     EXPECT_GF_NEAR(gtau, gtau3, sigma);
   }
+
+  // BlockGf test
+  auto G2 = gtau;
+  auto G_vec = std::vector{gtau, G2};
+  auto Bgtau = block_gf{G_vec};
```
Comment (Wentzell) [inline test/c++/gfs/gf_dlr.cpp]:
```suggestion
  auto Bgtau = block_gf{std::vector{gtau, gtau}};
```

File: test/c++/gfs/gf_dlr.cpp [hunk 2]
```diff
@@ -375,6 +384,16 @@ TEST(Gf, DLR_multivar_fit) {
   auto gtau2 = make_gf_imtime<0>(make_gf_imtime<1>(gcoef, n_tau), n_tau);
   EXPECT_GF_NEAR(gtau, gtau2);
 
+  // BlockGf
+  auto G2 = gtau;
+  auto G_vec = std::vector{gtau, G2};
+  auto Bgtau = block_gf{G_vec};
```
Comment (Wentzell) [inline test/c++/gfs/gf_dlr.cpp]:
```suggestion
  auto Bgtau = block_gf{std::vector{gtau, gtau}};
```

File: test/c++/gfs/gf_dlr.cpp [hunk 3]
```diff
@@ -484,11 +503,21 @@ TEST(Gf, DLR_h5) {
   auto gt = gf<dlr_imtime, scalar_valued>{{beta, Fermion, w_max, eps}};
   gt[tau_] << nda::clef::exp(-omega * tau_) / (1 + nda::clef::exp(-beta * omega));
   auto gc = make_gf_dlr(gt);
-  auto gw = make_gf_dlr_imtime(gc);
+  auto gw = make_gf_dlr_imfreq(gc);
+
+  // BlockGf test
+  auto G2 = gt;
+  auto G_vec = std::vector{gt, G2};
+  auto Bgtau = block_gf{G_vec};
```
Comment (Wentzell) [inline test/c++/gfs/gf_dlr.cpp]:
```suggestion
  auto Bgtau = block_gf{std::vector{gt, gt}};
```

File: test/c++/gfs/gf_dlr.cpp [hunk 4]
```diff
@@ -484,11 +503,21 @@ TEST(Gf, DLR_h5) {
...
+  auto Bgcoef = make_gf_dlr(Bgtau);
+  auto Bgiwn = make_gf_dlr_imfreq(Bgcoef);
 
   rw_h5(gt, "g_dlr_imtime");
   rw_h5(gc, "g_dlr");
   rw_h5(gw, "g_dlr_imfreq");
+  rw_h5(Bgtau, "g_dlr_imtime");
+  rw_h5(Bgcoef, "g_dlr");
+  rw_h5(Bgiwn, "g_dlr_imfreq");
```
Comment (Wentzell) [inline test/c++/gfs/gf_dlr.cpp]:
```suggestion
  rw_h5(Bgtau, "bg_dlr_imtime");
  rw_h5(Bgcoef, "bg_dlr");
  rw_h5(Bgiwn, "bg_dlr_imfreq");
```

File: test/python/base/gf_dlr.py [hunk 5]
```diff
@@ -103,6 +103,11 @@ def test_dlr_gfs_density(self):
         ref = np.exp(-e * beta) / (1 + np.exp(-e * beta))
         np.testing.assert_almost_equal(density(g_c), ref)
 
+        # block gf test
+        Bgw = BlockGf(name_list=['1', '2'], block_list=[g_c, g_c], make_copies=True)
+        for block, gf in Bgw:
+            np.testing.assert_almost_equal(density(g_c), ref)
```
Comment (Wentzell) [inline test/python/base/gf_dlr.py]: This is not really testing the any of the dlr BlockGf functions?

Also
```suggestion
        Bgw = BlockGf(name_list=['1', '2'], block_list=[g_c, g_c], make_copies=True)
        for block, gw in Bgw:
            np.testing.assert_almost_equal(density(gw), ref)
```

File: test/python/base/gf_dlr.py [hunk 6]
```diff
@@ -131,10 +136,16 @@ def test_dlr_gfs_imtime_interp(self):
...
 
+        # block gf test
+        Bgdlr = BlockGf(name_list=['1', '2'], block_list=[g, g], make_copies=True)
+
         for tau in np.linspace(0, beta, num=10):
             ref = np.sum(g.data * -np.exp(-rf * tau/beta) / ( 1 + np.exp(-rf) ))
             np.testing.assert_almost_equal(g(tau), ref)
 
+            for block, gf in Bgdlr:
+                np.testing.assert_almost_equal(Bgdlr[block](tau), ref)
```
Comment (Wentzell) [inline test/python/base/gf_dlr.py]:
```suggestion
            for block, gdlr in Bgdlr:
                np.testing.assert_almost_equal(gdlr(tau), ref)
```

**Review by Wentzell — CHANGES_REQUESTED** [review]
Thank you @the-hampel for the useful generalization and the fix.
See comments below

File: test/python/base/gf_dlr.py [hunk 7]
```diff
@@ -105,8 +117,9 @@ def test_dlr_gfs_density(self):
 
         # block gf test
         Bgw = BlockGf(name_list=['1', '2'], block_list=[g_c, g_c], make_copies=True)
+        Bden = Bgw.density()
         for block, gw in Bgw:
-            np.testing.assert_almost_equal(density(gw), ref)
+            np.testing.assert_almost_equal(Bden[block], ref)
```
Comment (Wentzell) [inline test/python/base/gf_dlr.py]:
```suggestion
            np.testing.assert_almost_equal(density(gw), ref)
```

File: test/python/base/gf_dlr.py [hunk 8]
```diff
@@ -86,7 +86,19 @@ def test_dlr_gfs_conversion(self):
...
+        Bg_c = make_gf_dlr(Bg_w)
+        Bg_t = make_gf_dlr_imtime(Bg_c)
+        for t in g_t.mesh:
+            ref = -np.exp(-e * t.value) / ( 1 + np.exp(-e * beta) )
+            np.testing.assert_almost_equal(Bg_c['1'](t), ref)
+
+        Bg_w_new = make_gf_dlr_imfreq(Bg_c)
+        for w in g_w_new.mesh:
+            np.testing.assert_almost_equal(Bg_w['1'][w], Bg_w_new['1'][w])
+            np.testing.assert_almost_equal(Bg_w['2'][w], Bg_w_new['2'][w])
```
Comment (Wentzell) [inline test/python/base/gf_dlr.py]:
```suggestion

        # BlockGf test
        Bg_w = BlockGf(block_list=[g_w, g_w], make_copies=True)
        Bg_c = make_gf_dlr(Bg_w)
        Bg_t = make_gf_dlr_imtime(Bg_c)
        for t in g_t.mesh:
            ref = -np.exp(-e * t.value) / ( 1 + np.exp(-e * beta) )
            np.testing.assert_almost_equal(Bg_c['0'](t), ref)

        Bg_w_new = make_gf_dlr_imfreq(Bg_c)
        assert_block_gfs_are_close(Bg_w, Bg_w_new)
```

File: test/python/base/gf_dlr.py (hunk 7 above)
Comment (the-hampel) [reply test/python/base/gf_dlr.py]: but that would defeat the purpose here. I wanted to call the density function first for the BlockGf in line 120 and the compare the block density array. Does this make sense?

Comment (Wentzell) [reply test/python/base/gf_dlr.py]: You are correct, I meant to suggest something along the lines of
```python
     for block, den in Bden:
            np.testing.assert_almost_equal(den, ref)
```

Comment (the-hampel) [reply test/python/base/gf_dlr.py]: ah yes that makes sense. I just updated the lines accordingly.

File: test/python/base/gf_dlr.py [hunk 9]
```diff
@@ -103,6 +113,11 @@ def test_dlr_gfs_density(self):
         ref = np.exp(-e * beta) / (1 + np.exp(-e * beta))
         np.testing.assert_almost_equal(density(g_c), ref)
 
+        # block gf test
+        Bgw = BlockGf(name_list=['1', '2'], block_list=[g_c, g_c], make_copies=True)
```
Comment (Wentzell) [inline test/python/base/gf_dlr.py]:
```suggestion
        Bgw = BlockGf(block_list=[g_c, g_c], make_copies=True)
```

File: test/python/base/gf_dlr.py [hunk 10]
```diff
@@ -131,9 +146,14 @@ def test_dlr_gfs_imtime_interp(self):
         g = Gf(mesh=m, target_shape=[])
         g.data[:] = np.random.randn(len(m))
 
+        # block gf test
+        Bgdlr = BlockGf(name_list=['1', '2'], block_list=[g, g], make_copies=True)
```
Comment (Wentzell) [inline test/python/base/gf_dlr.py]:
```suggestion
        Bgdlr = BlockGf(block_list=[g, g], make_copies=True)
```

**Pull request merged**